### ISSUE TITLE
Division Learning Example (help needed)

### ISSUE THREAD

**Issue by Philipp114** [nengo/nengo#1075]
Hi nengo users

Since my last question a few weeks ago i made quite some progress in different areas of how to implement some kind of calculator.
Now i am trying to have an Ensemble learn specific division values, but i am having trouble with the disconnection of the learning after some time.

Here is a little example which is very close to the given square example in the Tutorials:
The Program is supposed to learn to divide 10 by values between 2 and 5 and stop learning after some time to compute a given number.

```python
    import numpy as np

    import nengo
    from nengo.solvers import LstsqL2

    model = nengo.Network()
    with model:
    
	#creating values between 2 and 5
    input_node = nengo.Node(lambda t:1.5* np.sin(t)+3.5 if t<15 else 5)
	#dividing 10 by values between 2 and 5
    correct_div = nengo.Node(lambda t:10/(1.5* np.sin(t)+3.5))
    
 
    A = nengo.Ensemble(400, dimensions=1, radius=5)
    A_inverted = nengo.Ensemble(400, dimensions=1, radius=5)
    error = nengo.Ensemble(400, dimensions=1, radius=5)
    
    #learning taken from square example
    pts = 10/(1.5*np.sin(np.linspace(0, 2*np.pi, 400))+3.5)
    pts.shape = 400,1

    conn = nengo.Connection(A, A_inverted, eval_points=pts/A.radius)
    conn.solver = LstsqL2(weights=True)

    error_conn = nengo.Connection(error, A_inverted, modulatory=True)
    conn.learning_rule_type = nengo.PES(error_conn, learning_rate=3e-6)

	
    nengo.Connection(A_inverted, error, transform=-1)
	# I used the correct div as an external source, is this correct?
    nengo.Connection(correct_div, error, function=lambda x:x) 
    nengo.Connection(input_node, A)
	
	#This inhibit node is the main part of my confusion
	#Is it not possible to just cut a learning connection after a while?
	
    def inhibit(t):
        return 1.0 if t > 10.0 else 0.0

    inhib = nengo.Node(inhibit)
    nengo.Connection(inhib, error.neurons, transform=[[-1]] * error.n_neurons)
```

The first part of learning the correct values using the error seems to work just fine, but as soon as i try to disconnect the learning, it does not actually do so and keeps copying the correct div function even though it shouldnt anymore.
Since i really didnt change much from the example i am wondering where i have gone wrong...

Thank you in advance :D

**Comment by tcstewar** [comment]
Hello,

It looks to me like you just need stronger inhibition!  The scaling on the inhibition is such that a value of 1 is enough inhibition to take a neuron that's representing a 1 and push it down to represent a 0.  In your case, you have a radius of 5, so if the neurons are representing values larger than 1, they're not being inhibited enough.  I usually use something around 2 times the radius, so I think if you change your last line to 

```
nengo.Connection(inhib, error.neurons, transform=[[-10]] * error.n_neurons)
```

thin it should work.

**Comment by Philipp114** [comment]
Ahh i understand now, thank you for the quick response.

It does in fact help with inhibiting the signal, but sadly now it seems that the Ensemble has not learned the division at all... 
Does the type of function used for learning matter a lot? 

It seems so simple in the Example used for squaring.

**Comment by tcstewar** [comment]
This is a harder function to learn, and it looks like you're running into the standard "catastrophic forgetting" problem in neural networks.  Basically, it's doing a very good job of learning what to do given its current input, but those weight changes are destroying its ability to handle inputs that it has seen in the past.

The two standard ways of dealing with that are to a) decrease the learning rate, and b) vary the input more.  Here's what I tried:

```python
    input_node = nengo.Node(lambda t:(1.5*np.sin(10*t)+3.5) if t<15 else 5)
    
    #dividing 10 by values between 2 and 5
    correct_div = nengo.Node(lambda t:10/(1.5* np.sin(10*t)+3.5))

    ...

    conn.learning_rule_type = nengo.PES(error_conn, learning_rate=1e-6)
```

(so I sped up the sine wave to change 10 times faster, and I decreased the learning rate to 1e-6)

That seems to work for me!  Let me know if it works for you...

**Comment by arvoelke** [comment]
> The two standard ways of dealing with that are to a) decrease the learning rate, and b) vary the input more. 

in terms of non-standard ways, adding sparsity to the representation can also help with the problem of decoders overwriting eachother (since only the decoders of the active neurons will move). This is the approach taken when I learn my heteroassociative memory. However, that's not so relevant here since the ensemble is one-dimensional (it's not possible in one dimension to simultaneously encode the entire space while making the LIF tuning curves selective to only small portions of the space). 

Just thought I'd mention it in case more dimensions are added in the future. Also, as a recreational hack, you could artificially increase the dimensionality by mapping the input into some higher dimensional space (say the surface of a 2D-circle) and then play with the intercepts in the higher dimension (setting `intercepts=nengo.dists.Uniform(0.5, 1.0)` for the circle will make each neuron selective to a small piece of the pie -- i.e. a segment of the one-dimensional line),

**Comment by Philipp114** [comment]
> (so I sped up the sine wave to change 10 times faster, and I decreased the learning rate to 1e-6)

> That seems to work for me! Let me know if it works for you...

Yes it does work, thank you very much for this insight. So the decoder learning actually needs a certain timing.


Also thanks a lot for the information about 2Dimensional arrays since my aim is indeed a system with more dimensions. Just finding information about such specific problems has proven difficult.

**Comment by tcstewar** [comment]
> So the decoder learning actually needs a certain timing.

Pretty much all learning methods in general end up running into this problem.  :)

**Comment by jgosmann** [comment]
There haven't been any posts in a while here, so I am closing this issue. If you have further questions, @Philipp114, feel free to ask them in the new [Nengo forum](https://forum.nengo.ai/).
